### PR TITLE
Protobuf 3.0

### PR DESCRIPTION
**Pull request by apjanke**
Protobuf major version upgrade to 3.x.

This is an updated version of #3415, preserving original author credit.
- Reorders and squashes a couple commits for a tidy commit history
- Incorporates a compatibility fix for zbackup and removed the bogus version-specific dependency

Closes #3415 
Closes #3803 
## Stuff still broken
- [X] nanopb-generator is broken
- [ ] upscaledb is failing in the test; looks like it's crashing

### CODE REVIEW

**Comment by apjanke** [comment]
Failed build:

https://bot.brew.sh/job/Homebrew%20Core%20Pull%20Requests/6660/

* nanopb-generator broken
* ola needed revision bump (now done)
* upscaledb breaking in test
* protobuf-swift audit failure (now fixed)

**Comment by apjanke** [comment]
@agatti Any chance you know enough about `nanopb-generator` to get it working with `protobuf` 3.0?

File: Formula/supersonic.rb [hunk 1]
```diff
@@ -3,14 +3,9 @@ class Supersonic < Formula
...
+  revision 1
 
-  bottle do
-    revision 1
-    sha256 "cd76e23c748ec6cfb520800a1890e691680921efa4e6e1bf1dd991b0557e6d33" => :el_capitan
-    sha256 "75c8903d7d637aa495c83eb3a3569b627fbdf9799e907a7d0fff536cf3ddb155" => :yosemite
-    sha256 "c72cb21c7f1efc5a790be376e55dcc4a1edc76cda6686c2ba7d3d7f8c2937321" => :mavericks
-    sha256 "2869dba7bb685f7f6e4df504b87e01a7dda685afc3f27fa1c6010c150982317f" => :mountain_lion
-  end
+  depends_on "gflags" => :linked
```
Comment (DomT4) [inline Formula/supersonic.rb]: This sort of syntax is dead now. Will need to remove it.

Comment (apjanke) [inline Formula/supersonic.rb]: Fixed.

**Comment by agatti** [comment]
@apjanke I can take a look later this evening, it probably is a matter of infrastructure setup.  From a quick look at the log, it seems to be about https://pypi.python.org/pypi/six not being installed in the environment, and the test being ran without specifying the protobuf format version - so it'll default to v2.

**Comment by ilovezfs** [comment]
@BrewTestBot test this please

**Comment by apjanke** [comment]
Thanks @agatti.

FYI, in #4148 the `nanopb-generator` test was modified to explicitly specify the protocol version as `proto2`. This PR has since been rebased on master, so it includes that format version specification.

Maybe `six` needs to be added as a vendored resource dependency inside `nanopb-generator`?

**Comment by ilovezfs** [comment]
@apjanke I just bumped libphonenumber so this PR will need a little kick.

**Comment by apjanke** [comment]
Rebased on master to kick it.

**Comment by DomT4** [comment]
@BrewTestBot test this please

**Comment by DomT4** [comment]
The fetch failures should be resolved now, at least.

**Comment by Noctem** [comment]
You're welcome to cherry-pick or merge Noctem/homebrew-core@87f632ce327e13500c16e5a40e4b613f33d4c770 if you'd like updated Python modules.

**Comment by DomT4** [comment]
Probably not a bad idea. Looks like we're missing a `six` resource as well, which is causing the `nanopb-generator` test to fail, apparently.

**Comment by apjanke** [comment]
Thanks @Noctem! That looks like it will solve some of our problems here. Gonna cherry-pick that in to this PR. The `six` there looks like it may work for `nanopb-generator` as well.

**Comment by apjanke** [comment]
Okay, that fixed `nanopb-generator` for me locally, leaving only `upscaledb` broken.

Upscaledb vendors an older series of `libuv`, version 0.10.36. Even though libuv is [up to 1.9.x now](https://github.com/libuv/libuv/releases), they just released a 0.10.37 patch version back in June, about the time Protobuf 3.0 came out. Bumping the vendored `libuv` to 0.10.37 fixed the `brew test upscaledb` crash for me locally.

Amended the PR with these. Think it might work now.

**Comment by apjanke** [comment]
@BrewTestBot test this please

**Comment by apjanke** [comment]
Darnit. The upscaledb test is [still failing on the test bot](https://bot.brew.sh/job/Homebrew%20Core%20Pull%20Requests/6988/), even though it works for me locally.

Created a separate upscaledb libuv bump at https://github.com/Homebrew/homebrew-core/pull/4389 to see if this is actually a regression with protobuf 3.0, or if it's just exposing an existing error.

**Comment by jmhodges** [comment]
Anything I can do to help with this?

**Comment by ilovezfs** [comment]
@apjanke do you mind rebasing this? I think we should merge it regardless of the upscaledb issue and then a separate PR can deal with that.

**Comment by apjanke** [comment]
Rebased on `master`.

I'm also inclined to merge this as is, though that may just be my eagerness for new releases talking. It seems likely that merging this will break our `upscaledb`, and I don't know how many users that will affect.

@jmhodges The only blocker on this is the failing `upscaledb` test. If you know why upscaledb is crashing with the new `protobuf`, that could help clear this up.

**Comment by ilovezfs** [comment]
@apjanke I see this as a fairly straightforward case of "perfect is the enemy of the good" and think in general these PRs with several reverse dependencies should not be blocked by the few outliers (say like < 20% of the reverse dependencies) that will be broken, even if we let those outliers lie broken for a bit rather than immediately exiling them to the boneyard as part of the breaking PR, so long as there are PRs opened subsequently that show what broke, and we report the problems to the respective upstreams as soon as it's clear we're not going to fix it ourselves.

The alternative is being stuck with an old version in core, which is almost, though not exactly, the same as having a broken version in core, since it begins to impede people's work in just about the same way something broken would.

**Comment by jmhodges** [comment]
I would debug from the previously failing build, but it looks like its 404ing! Working out how to test this locally. Don't see docs immediately. If I don't figure it out, I'll wait for this current build to fail.

**Comment by jmhodges** [comment]
Ah, okay, I somehow need to get the `brew linkage` command going. I see its in dev-cmd but no docs on how to get it or the rest of those installed.

**Comment by ilovezfs** [comment]
it's installed. Library/Homebrew/dev-cmd ... just export HOMEBREW_DEVELOPER=1 and voilà

**Comment by jmhodges** [comment]
Thanks!

**Comment by jmhodges** [comment]
Okay, yep, works locally. Great.

**Comment by ilovezfs** [comment]
@jmhodges also for more toys, `brew tap homebrew/dev-tools`

**Comment by jmhodges** [comment]
Oh, could it be the created bottle doesn't have it? We uninstall protobuf, and then try the new bottle and then tests start failing.

==> brew uninstall --force protobuf
==> brew uninstall --force autoconf automake libtool xz
==> brew install ./protobuf-3.0.0.el_capitan.bottle.tar.gz
==> brew test protobuf --verbose

**Comment by jmhodges** [comment]
Boom! Reproduced by installing from the bottle!

Running

==> brew install --build-bottle --verbose protobuf
==> brew audit protobuf
==> brew bottle --verbose --json protobuf
==> brew bottle --merge --write --no-commit ./protobuf-3.0.0.el_capitan.bottle.json

and then

    brew linkage upscaledb --test
    Missing libraries:
      /usr/local/opt/protobuf/lib/libprotobuf.9.dylib

**Comment by jmhodges** [comment]
I'm at my limits tonight. Have jury duty early in the morning. Perhaps someone with some knowledge of bottling can understand why that lib might not be copied into the bottle.

**Comment by ilovezfs** [comment]
it segfaults though ...

**Comment by jmhodges** [comment]
What segfault? I don't see a segfault in the current build failure.

**Comment by ilovezfs** [comment]
https://gist.github.com/ilovezfs/6c899fa5e344b92e165d191d73867356

**Comment by jmhodges** [comment]
How are you getting that? It seems plausible a test could segfault if it was expecting to dlopen a dylib or some such and didn't check its errors.

The dylib no longer exists on my local machine when I install from the newly made protobuf bottle. That seems to be the first failure. We can look at the upscaledb test failure after that.

**Comment by jmhodges** [comment]
Oh interesting! The bottle tar.gz has a `libprotobuf.10.dylib`. Wonder how it got that!

**Comment by ilovezfs** [comment]
```
iMac-TMP:~ joe$ brew linkage --test protobuf
No broken dylib links
iMac-TMP:~ joe$ brew linkage --test upscaledb
No broken dylib links
iMac-TMP:~ joe$ brew linkage upscaledb
System libraries:
  /System/Library/Frameworks/ApplicationServices.framework/Versions/A/ApplicationServices
  /System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation
  /System/Library/Frameworks/CoreServices.framework/Versions/A/CoreServices
  /System/Library/Frameworks/Foundation.framework/Versions/C/Foundation
  /usr/lib/libSystem.B.dylib
  /usr/lib/libc++.1.dylib
  /usr/lib/libz.1.dylib
Homebrew libraries:
  /usr/local/opt/boost/lib/libboost_chrono-mt.dylib (boost)
  /usr/local/opt/boost/lib/libboost_filesystem-mt.dylib (boost)
  /usr/local/opt/boost/lib/libboost_system-mt.dylib (boost)
  /usr/local/opt/boost/lib/libboost_thread-mt.dylib (boost)
  /usr/local/opt/openssl/lib/libcrypto.1.0.0.dylib (openssl)
  /usr/local/opt/protobuf/lib/libprotobuf.10.dylib (protobuf)
  /usr/local/Cellar/upscaledb/2.2.0_1/lib/libupscaledb.0.dylib (upscaledb)
  /usr/local/Cellar/upscaledb/2.2.0_1/lib/libupsserver.0.dylib (upscaledb)
  /usr/local/Cellar/upscaledb/2.2.0_1/libexec/libuv/lib/libuv.dylib (upscaledb)
iMac-TMP:~ joe$ mdutil -t /usr/local/opt/protobuf/lib/libprotobuf.10.dylib
/usr/local/Cellar/protobuf/3.0.0/lib/libprotobuf.10.dylib
iMac-TMP:~ joe$ ls -l /usr/local/Cellar/protobuf/3.0.0/lib/libprotobuf.10.dylib
-r--r--r--  1 joe  admin  2143968 Sep  7 00:43 /usr/local/Cellar/protobuf/3.0.0/lib/libprotobuf.10.dylib
iMac-TMP:~ joe$
```
I'm just pulling this PR `brew pull 4145` and then brew install -s protobuf ; brew install -s upscaledb ; brew test -vd upscaledb ; scp test joe@10.0.1.15: ; ctrl D ; cd ~ ; sudo lldb ./test ; r`

**Comment by jmhodges** [comment]
Right, it seems that `install -s protobuf` isn't the right step. You've got to create and then install from the bottle. Otherwise, everything works.

**Comment by MikeMcQuaid** [comment]
@ilovezfs If we add a `revision` to `upscaledb` I'm almost 100% sure this will Just Work.

**Comment by jmhodges** [comment]
Well, the `linkage upscaledb --test`works, I mean.

**Comment by MikeMcQuaid** [comment]
Similarly with `gjstest`

**Comment by ilovezfs** [comment]
@MikeMcQuaid this is protobuf installed from source then upscaledb installed from source then segfault. but I can try again with a super clean environment if you all think something is amiss for me.

**Comment by MikeMcQuaid** [comment]
@ilovezfs I guess you've found another issue but those two definitely need a revision bump before pulling this PR.

**Comment by jmhodges** [comment]
Yeah, @MikeMcQuaid is on to something. This seems very much because the build-bot is not installing upscaledb from source when it runs its test, but using the new bottle that makes the `libprotobuf.10.dylib` instead of `libprotobuf.9.dylib`.

**Comment by ilovezfs** [comment]
@MikeMcQuaid yes, of course. I'm assuming apjanke left it out due to the segfault.

**Comment by jmhodges** [comment]
Oh, I expected to be able to use the build output to guide what needed debugging. Would love to have those revision bumps in to make sure we're all on the same page.

**Comment by ilovezfs** [comment]
the revision bumps are irrelevant if both are installed from source locally.

**Comment by jmhodges** [comment]
I think you misunderstand why I'm asking for it! When you're trying to help out or just coming back to a problem after a bit away from it, it's nice to be able to look at documented test failures and know they are the actual test failures people are trying to debug elsewhere.

It's a communication and clarity thing to others or future you. We just wasted a bunch of people's time because we didn't have those revision bumps in showing the exact test failure that was the problem. Plus, the PR wouldn't have ever made it in without them, so why not put them in and be that much closer to having this shipping?

**Comment by ilovezfs** [comment]
because I'm not apjanke so I can't do that otherwise I would for the reasons you stated

**Comment by jmhodges** [comment]
Ah, I've not been asking you to do it. I thought it was clear from context that I was hoping apjanke would.

**Comment by ilovezfs** [comment]
He will, no doubt, but given the hour in New York that's unlikely to happen imminently, I suspect.

**Comment by ilovezfs** [comment]
yup same segfault with a clean environment.

**Comment by jmhodges** [comment]
Hrm, tossing "-g" into the CFLAGS doesn't seem to be enough to get upstreamdb to build with symbols. Dropping the `--disable-debug` (which maybe is causing the issue) causes the build to break on a busted assert.

Really got to hit the bricks. But we're closer!

**Comment by ilovezfs** [comment]
that's due to https://github.com/Homebrew/brew/blob/d39eeac0d5041dbbc41f939ef2dbc02823c82fcd/Library/Homebrew/shims/super/cc#L152

you have to remove the  /^-g\d?$/

**Comment by ilovezfs** [comment]
@jmhodges with symbols:
```
iMac-TMP:~ joe$ sudo lldb ./test
Password:
(lldb) target create "./test"
Current executable set to './test' (x86_64).
(lldb) r
Process 66489 launched: './test' (x86_64)
Process 66489 stopped
* thread #1: tid = 0x1426125, 0x0000000100171d36 libupscaledb.0.dylib`::ups_env_create(henv=0x00007fff5fbffcb0, filename=<unavailable>, flags=0, mode=<unavailable>, param=0x0000000000000000) + 163 at upscaledb.cc:286, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x0)
    frame #0: 0x0000000100171d36 libupscaledb.0.dylib`::ups_env_create(henv=0x00007fff5fbffcb0, filename=<unavailable>, flags=0, mode=<unavailable>, param=0x0000000000000000) + 163 at upscaledb.cc:286
   283 	  *henv = 0;
   284 	
   285 	  /* creating a file in READ_ONLY mode? doesn't make sense */
-> 286 	  if (unlikely(isset(flags, UPS_READ_ONLY))) {
   287 	    ups_trace(("cannot create a file in read-only mode"));
   288 	    return (UPS_INV_PARAMETER);
   289 	  }
```

**Comment by jmhodges** [comment]
Shoot, I'm not able to get debug symbols. Added a  `ENV.prepend "CFLAGS", "-g"` to upscaledb's install method, commented out that entire `when` clause in `shims/super/cc` after removing the `-g` regex bit didn't do it, and even got a patch working to allow the `--disable-debug` to be dropped. But no luck.

Maybe there's a copy of cc somewhere I'm supposed to edit, instead?

**Comment by ilovezfs** [comment]
@jmhodges try `--env=std` instead

**Comment by jmhodges** [comment]
Nope, that didn't do it either.

    brew uninstall upscaledb; brew install --env=std -s --debug  upscaledb  && brew test -vd upscaledb

**Comment by ilovezfs** [comment]
--env=std and these in the formula
```
      ENV.prepend "CFLAGS", "-g"
      ENV.prepend "CXXFLAGS", "-g"
```

**Comment by jmhodges** [comment]
Ah! CXXFLAGS. Duh.

**Comment by jmhodges** [comment]
Yeah, that didn't do it either. I've been running the `brew test -vd upscaledb`, letting it fail, then hopping into the tmp directory where its binary is and running it under `lldb test` and getting nada for symbols. After jury duty, I'll take another whack in a different direction.

**Comment by jmhodges** [comment]
(Well, to be clear, I get symbols from the test binary, but not the upscaledb lib)

**Comment by jmhodges** [comment]
Okay, yeah, I'm going to need some different help because I just cannot get the upstreamdb lib to build with debug symbols on my local machine. There's been a lot of undocumented stuff so far in this thread, and I'm assuming it's just another one of those little tweaks.

**Comment by ilovezfs** [comment]
@jmhodges how do you usually build it? You should be able to just point CPPFLAGS and LDFLAGS at the installed protobuf3 and build however you usually do outside of brew.

**Comment by jmhodges** [comment]
I've never built it before. I'm doing this to get a stable version of protobuf3 without breaking people's builds

**Comment by jmhodges** [comment]
Oh, and using homebrew, the commands I'm using are listed above in exactness

**Comment by ilovezfs** [comment]
@jmhodges Ah, OK. Thanks for clarifying.

**Comment by ilovezfs** [comment]
@jmhodges I fixed it :) ...
```
perl -p -i -e 's/isset/myisset/g' `grep -ril isset *`
```

:smiling_imp:

**Comment by jmhodges** [comment]
Great! I'll make a patch. In for the night, but soon, if @apjanke doesn't beat me to it.

**Comment by ilovezfs** [comment]
@jmhodges @apjanke narrowed it down further and this does the job:
```
    inreplace "./src/5upscaledb/upscaledb.cc",
      "#  include \"2protobuf/protocol.h\"",
      "#  include \"2protobuf/protocol.h\"\n#define isset(f, b)       (((f) & (b)) == (b))"
```
or as a patch
```
diff --git a/src/5upscaledb/upscaledb.cc b/src/5upscaledb/upscaledb.cc
index 30496d0..8ada3aa 100644
--- a/src/5upscaledb/upscaledb.cc
+++ b/src/5upscaledb/upscaledb.cc
@@ -32,6 +32,7 @@
 #include "2page/page.h"
 #ifdef UPS_ENABLE_REMOTE
 #  include "2protobuf/protocol.h"
+#define isset(f, b)       (((f) & (b)) == (b))
 #endif
 #include "2compressor/compressor_factory.h"
 #include "2device/device.h"
```

Note that if you move that define above `#  include "2protobuf/protocol.h"` then it will still segfault. Putting back the original definition of `isset` from `0root/root.h` after that include fixes it. I'm not sure exactly what in protocol.h causes the redefinition but I think it's something in the generated header named `"2protobuf/messages.pb.h"` which itself includes some other files.

**Comment by ilovezfs** [comment]
OK here's the offender: https://github.com/google/protobuf/blob/master/src/google/protobuf/stubs/port.h#L62

and macOS's `<sys/param.h>` defines
```
/* Bit map related macros. */
#define setbit(a,i)     (((char *)(a))[(i)/NBBY] |= 1<<((i)%NBBY))
#define clrbit(a,i)     (((char *)(a))[(i)/NBBY] &= ~(1<<((i)%NBBY)))
#define isset(a,i)      (((char *)(a))[(i)/NBBY] & (1<<((i)%NBBY)))
#define isclr(a,i)      ((((char *)(a))[(i)/NBBY] & (1<<((i)%NBBY))) == 0)
```
port.h is included by
https://github.com/google/protobuf/blob/master/src/google/protobuf/stubs/common.h#L40
which is included by
2protobuf/messages.pb.h
which is the generated header included by
2protobuf/protocol.h
which is included by
src/5upscaledb/upscaledb.cc after it has included 0root/root.h, where the upscaledb definition of isset occurs.

So renaming isset everywhere in upscaledb may indeed be the right answer here so that this collision never happens again.

**Comment by apjanke** [comment]
Nice detective work!

Reported upstream: https://github.com/cruppstahl/upscaledb/issues/82

I think the simple `inreplace` or patch may be the way to go here: that'll take care of it for this version of upscaledb, and we can expect upstream to deal with it in a future release. Less work and code to read than renaming it everywhere. (Unless `inreplace` supports recursive globs?)

I've updated this PR with the workaround. If this build is green, I think we're good to merge.

**Comment by apjanke** [comment]
@ilovezfs said:
> @apjanke I see this as a fairly straightforward case of "perfect is the enemy of the good" and think in general these PRs with several reverse dependencies should not be blocked by the few outliers (say like < 20% of the reverse dependencies) that will be broken, even if we let those outliers lie broken for a bit rather than immediately exiling them to the boneyard...

I see where you're coming from here, and I would also not like to block major version upgrades on a few less-used or more slowly developed formulae. But since we're talking about breaking a previously-shipped formula, I don't think "perfect" is the term for it; more like "not-broken". I'm real hesitant to push out stuff that breaks any formula.

I'd rather deal with situations like this by relaxing the "one version in core" guideline, and doing something like bumping protobuf to 3.0 while grandfathering in a `protobuf2` formula for back-compatibility for packages that need it (either in core, or by allowing a dep on `homebrew/versions`). Not every package has a lot of developer resources, and breaking formula that don't switch quickly to the latest major version of a library seems like punishing the developers and users of packages from smaller projects, or those that have a more conservative or quality-oriented development process.

**Comment by ilovezfs** [comment]
>Unless inreplace supports recursive globs?

It does. The first argument can be an array.

**Comment by ilovezfs** [comment]
@apjanke we can defer the must-not-break-anything/multiple-versions/who-gets-to-suffer conversation since it's now moot with respect to this PR.

**Comment by apjanke** [comment]
Build failed due to `gjstest` linkage problem.

https://bot.brew.sh/job/Homebrew%20Core%20Pull%20Requests/7445/

Adding a revision bump to address.

**Comment by jmhodges** [comment]
Thanks to both you!

**Comment by ilovezfs** [comment]
@apjanke 
```
iMac-TMP:homebrew-core joe$ brew pull https://github.com/Homebrew/homebrew-core/pull/4145
==> Fetching patch 
Patch: https://github.com/Homebrew/homebrew-core/pull/4145.patch
==> Applying patch
Applying: protobuf 3.0.0
Applying: bump revisions for protobuf 3.0
Using index info to reconstruct a base tree...
M	Formula/libphonenumber.rb
Falling back to patching base and 3-way merge...
Auto-merging Formula/libphonenumber.rb
CONFLICT (content): Merge conflict in Formula/libphonenumber.rb
error: Failed to merge in the changes.
Patch failed at 0002 bump revisions for protobuf 3.0
The copy of the patch that failed is found in: .git/rebase-apply/patch
When you have resolved this problem, run "git am --continue".
If you prefer to skip this patch, run "git am --skip" instead.
To restore the original branch and stop patching, run "git am --abort".
Error: Patch failed to apply: aborted.
```
lol

**Comment by apjanke** [comment]
*sigh*; rebases again...

**Comment by apjanke** [comment]
finally, a green light! Merging if no objections.

**Comment by MikeMcQuaid** [comment]
@apjanke :shipit:

**Comment by ilovezfs** [comment]
:boat: :rocket:

**Comment by DomT4** [comment]
The rebase gunked up a bit 😢. Check out `libphonenumber`, and `nanopb-generator` seems to have been double revision bumped.

**Comment by apjanke** [comment]
Thanks for catching that. I've amended to fix. libphonenumber is cleaned up. nanopb-generator had been revision bumped twice (once in the original PR, and once in a subsequent fix) I've reordered and cleaned up the commits so it has a single revision bump and its commit history is cleaner. Also tweaked the original commit to not remove bottle stanzas, for a cleaner commit history.

**Comment by DomT4** [comment]
Some of the `revision`s landed in the wrong place that time, but let's get this shipped 👍.

The revision placement isn't problematic more than stylistically, so we can fix that in post rather than torturing the CI a bit more.

**Comment by apjanke** [comment]
Yeah, I'll clean that up in post.

Merged! @ilya-medvedev, thank you for the original PR!

Let me know if there are any issues.

**Comment by jmhodges** [comment]
Y'all are great! Thank you so much, @apjanke!

**Pull request closed without merging**